compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -1123,7 +1123,7 @@ pub(super) fn transmute_immediate<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     // While optimizations will remove no-op transmutes, they might still be
     // there in debug or things that aren't no-op in MIR because they change
     // the Rust type but not the underlying layout/niche.
-    if from_scalar == to_scalar {
+    if from_scalar == to_scalar && from_backend_ty == to_backend_ty {
```

Comment:
Maybe we should add some more comments on why both of these must obtain

Reply:
The entire function is cursed. It talks about transmuting "immediates" but then takes `Scalar` information as arguments. This needs some refactor figuring out which class of types it is meant for, and aligning name and signature and comments.